### ISSUE TITLE
Incorrect ArgumentError: wrong number of arguments

### ISSUE THREAD

**Issue by notEthan** [jruby/jruby#8382]
JRuby seems to get confused about method arity in some situation involving blocks and Enumerator.
I encountered this and have pared the code that triggers it down to this:

```ruby
class BlockHolder
  def initialize(block)
    @block = block
  end

  def call(*a, **kw)
    @block.call(*a, **kw)
  end
end

class S
  def initialize
    to_enum(:call_via_blockholder).to_a
  end

  def call_via_blockholder(&block)
    BlockHolder.new(block).call
  end
end

def passkw(kw: )
end

passkw(kw: S.new)
```

* JRuby version: jruby 9.4.8.0 (3.1.4) 2024-07-02 4d41e55a67 OpenJDK 64-Bit Server VM 17.0.5+8 on 17.0.5+8 +jit [x86_64-darwin]
* also reproduces on 9.4.2.0
* Operating system: consistent on macos and ubuntu

**Expected Behavior**: no error (passes on mri, truffleruby)
**Actual Behavior**: ArgumentError: wrong number of arguments (given 1, expected 0)
  passkw at test.rb:21
  <main> at test.rb:24

**Comment by headius** [comment]
This is indeed a bug in how we handle delegating keyword arguments through a `to_enum` Enumerator.

**Comment by enebo** [comment]
Just a note in case I don't look into this one.  The issue is almost entirely due to us clearing TC.callInfo when the native Java method is calling another method which still needs that callInfo intact.  It should just be tracing to the right @JRubyMethod and then making sure a) keywords = true b) we do not clear callInfo (which could possible be us restoring it if there is some other external call like a coercion protocol call).

**Comment by headius** [comment]
Very likely that there's Enumerator methods in Java that are not propagating the callInfo through when redispatching to the block, so hopefully this won't be too tricky to fix.

**Comment by headius** [comment]
The issue was that RubyEnumerable's Java-based block impls for `to_a`, `to_h`, and `tally` did not clear the `callInfo` for kwarg info, potentially leaving the bit marking `CALL_KEYWORD_EMPTY` set for a subsequent call (in this case `passkw`). I've fixed all three and added internal specs for the behavior.

RubyEnumerable's block for `chunk` also fails to clear the bit, but as it immediately does a no-argument call to `each` it is not possible to observe the bug directly. For sanitary reasons I also modified the `chunk` block to clear `callInfo`.

Fixed by #8389.
